---
title: "human exacerbation data"
author: "Mingming Zhang & Amrit Singh"
date: "December 22, 2025"
output:
  html_document:
    toc: true
    toc_depth: 3
    toc_float: true
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, cache = FALSE, warning = FALSE, message = FALSE)
```

## Setup: Libraries and utilities

```{r, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

library(anndata)
library(Seurat)
library(tidyverse)
library(SingleCellExperiment)
library(muscat)
library(scater)
library(limma)
library(ComplexHeatmap)
library(circlize)
library(here)
source(here::here("src", "utils", "sc_diffexp.R"))
latephase_exacer_biomarkers <- readRDS(here::here("results", "latephase_exacer_biomarkers.rds"))
```

## import single cell data

```{r}
data <- read_h5ad(here::here("data", "preprocessedData",  "GSE193816", "GSE193816_all_cells_data.h5ad"))
data <- CreateSeuratObject(counts = t(as.matrix(data$X)), 
                           meta.data = data$obs)

## mnp
mnp <- read_h5ad(here::here("data","preprocessedData",  "GSE193816", "GSE193816_mnp_data.h5ad"))
mnp <- CreateSeuratObject(counts = t(as.matrix(mnp$X)), 
                           meta.data = mnp$obs)

## t cells
tcell <- read_h5ad(here::here("data","preprocessedData",  "GSE193816", "GSE193816_t_cell_data.h5ad"))
tcell <- CreateSeuratObject(counts = t(as.matrix(tcell$X)), 
                          meta.data = tcell$obs)
```

## quality metrics

### All cells

```{r}
VlnPlot(data, features = c("nFeature_RNA", "nCount_RNA", 
                           "percent_mito"), ncol = 5)
```

### MNP cells

```{r}
VlnPlot(mnp, features = c("nFeature_RNA", "nCount_RNA", 
                           "percent_mito"), ncol = 5)
```

### T cells

```{r}
VlnPlot(tcell, features = c("nFeature_RNA", "nCount_RNA", 
                           "percent_mito"), ncol = 5)
```

## Dimension reduction plots

### All cells

```{r}
data <- NormalizeData(data,
                      normalization.method = "LogNormalize") %>% 
  FindVariableFeatures(selection.method = "vst", 
                       nfeatures = 2000) %>% 
  ScaleData(features = rownames(data)) %>% 
  RunPCA() %>% 
  RunUMAP(dims = 1:50, reduction.name = 'umap.rna', 
          reduction.key = 'rnaUMAP_')
Idents(data) <- data@meta.data$cluster

allcells <- DimPlot(data, reduction = "umap.rna", label = TRUE, 
                   label.size = 4, repel = TRUE, group.by = "cluster") + 
  ggtitle("RNA - all cells") +
  xlab("UMAP 1") +
  ylab("UMAP 2")
ggsave(here::here("results", "allcells_scrnaseq.png"), allcells, height = 4, width = 6)
allcells
```

##### save markers of all cells

```{r}
res <- Seurat::FindAllMarkers(data, only.pos = TRUE)
cell_markers <- res %>% 
  group_by(cluster) %>% 
  dplyr::slice(1:10)
cellmarkers <- split(cell_markers$gene, cell_markers$cluster)
saveRDS(cellmarkers, here::here("results", "cellmarkers_GSE193816.rds"))

cellmarkers
```

### MNP cells

```{r}
mnp <- NormalizeData(mnp,
                      normalization.method = "LogNormalize") %>% 
  FindVariableFeatures(selection.method = "vst", 
                       nfeatures = 2000) %>% 
  ScaleData(features = rownames(mnp)) %>% 
  RunPCA() %>% 
  RunUMAP(dims = 1:50, reduction.name = 'umap.rna', 
          reduction.key = 'rnaUMAP_')

mnpcells <- DimPlot(mnp, reduction = "umap.rna", label = TRUE, 
                   label.size = 2.5, repel = TRUE, group.by = "annotation") + 
  ggtitle("RNA - MNP cells") +
  xlab("UMAP 1") +
  ylab("UMAP 2")
ggsave(here::here("results", "mnpcells_scrnaseq.png"), mnpcells, height = 4, width = 6)
mnpcells
```

### T cells

```{r}
tcell <- NormalizeData(tcell,
                      normalization.method = "LogNormalize") %>% 
  FindVariableFeatures(selection.method = "vst", 
                       nfeatures = 2000) %>% 
  ScaleData(features = rownames(tcell)) %>% 
  RunPCA() %>% 
  RunUMAP(dims = 1:50, reduction.name = 'umap.rna', 
          reduction.key = 'rnaUMAP_')

tcells <- DimPlot(tcell, reduction = "umap.rna", label = TRUE, 
                   label.size = 2.5, repel = TRUE, group.by = "annotation") + 
  ggtitle("RNA - T cells") +
  xlab("UMAP 1") +
  ylab("UMAP 2")
ggsave(here::here("results", "tcells_scrnaseq.png"), tcells, height = 4, width = 6)
tcells
```

## Cell-type proportions

### All cells

```{r}
data@meta.data %>% 
  dplyr::select(id, group, condition, cluster) %>%
  dplyr::group_by(id, group, condition, cluster) %>% 
  mutate(n = n()) %>% 
  mutate(condition = factor(as.character(condition), 
                            levels = c("Bln", "Dil", "Ag"))) %>% 
  ggplot(aes(x = condition, y = n, color = group)) +
  geom_boxplot() +
  facet_wrap(~cluster, scales = "free") +
  theme_bw()
```

### MNP cells

```{r}
mnp@meta.data %>% 
  dplyr::select(id, group, condition, annotation) %>%
  dplyr::group_by(id, group, condition, annotation) %>% 
  mutate(n = n()) %>% 
  mutate(condition = factor(as.character(condition), 
                            levels = c("Bln", "Dil", "Ag"))) %>% 
  ggplot(aes(y = annotation, x = n, color = condition)) +
  geom_boxplot() +
  facet_grid(~group) +
  theme_bw()
```

### T cells

```{r}
tcell@meta.data %>% 
  dplyr::select(id, group, condition, annotation) %>%
  dplyr::group_by(id, group, condition, annotation) %>% 
  mutate(n = n()) %>% 
  mutate(condition = factor(as.character(condition), 
                            levels = c("Bln", "Dil", "Ag"))) %>% 
  ggplot(aes(y = annotation, x = n, color = condition)) +
  geom_boxplot() +
  facet_grid(~group) +
  theme_bw()
```

## Differential expression analysis

### All cells

```{r}
sct_counts <- GetAssayData(data, assay = "RNA", layer = "counts")
metadata_aa <- data@meta.data %>% 
  dplyr::filter(group == "AA",
                condition != "Dil")
metadata_aa$group_id <- droplevels(metadata_aa$condition)
metadata_aa$sample_id <- droplevels(metadata_aa$Channel)
all_aa = sc_diffexp(sct_counts, metadata_aa, cell_ann_col = "cluster")
all_aa$toptables$condition <- "AA"
all_aa$toptables$cc <- "All cells"
metadata_ac <- data@meta.data %>% 
  dplyr::filter(group == "AC",
                condition != "Dil")
metadata_ac$group_id <- droplevels(metadata_ac$condition)
metadata_ac$sample_id <- droplevels(metadata_ac$Channel)
all_ac = sc_diffexp(sct_counts, metadata_ac, cell_ann_col = "cluster")
all_ac$toptables$condition <- "AC"
all_ac$toptables$cc <- "All cells"

saveRDS(all_aa$toptables,
        here::here("results", "sc_exacer_aa_toptable.rds"))
```

### MNP cells

```{r}
sct_counts <- GetAssayData(mnp, assay = "RNA", layer = "counts")
metadata_aa <- mnp@meta.data %>% 
  dplyr::filter(group == "AA",
                condition != "Dil")
metadata_aa$group_id <- droplevels(metadata_aa$condition)
metadata_aa$sample_id <- droplevels(metadata_aa$Channel)
mnp_aa = sc_diffexp(sct_counts, metadata_aa, cell_ann_col = "annotation")
mnp_aa$toptables$condition <- "AA"
mnp_aa$toptables$cc <- "MNP cells"
metadata_ac <- mnp@meta.data %>% 
  dplyr::filter(group == "AC",
                condition != "Dil")
metadata_ac$group_id <- droplevels(metadata_ac$condition)
metadata_ac$sample_id <- droplevels(metadata_ac$Channel)
mnp_ac = sc_diffexp(sct_counts, metadata_ac, cell_ann_col = "annotation")
mnp_ac$toptables$condition <- "AC"
mnp_ac$toptables$cc <- "MNP cells"
```

### T cells

```{r}
sct_counts <- GetAssayData(tcell, assay = "RNA", layer = "counts")
metadata_aa <- tcell@meta.data %>% 
  dplyr::filter(group == "AA",
                condition != "Dil")
metadata_aa$group_id <- droplevels(metadata_aa$condition)
metadata_aa$sample_id <- droplevels(metadata_aa$Channel)
tcell_aa = sc_diffexp(sct_counts, metadata_aa, cell_ann_col = "annotation")
tcell_aa$toptables$condition <- "AA"
tcell_aa$toptables$cc <- "T cells"
metadata_ac <- tcell@meta.data %>% 
  dplyr::filter(group == "AC",
                condition != "Dil")
metadata_ac$group_id <- droplevels(metadata_ac$condition)
metadata_ac$sample_id <- droplevels(metadata_ac$Channel)
tcell_ac = sc_diffexp(sct_counts, metadata_ac, cell_ann_col = "annotation")
tcell_ac$toptables$condition <- "AC"
tcell_ac$toptables$cc <- "T cells"
```

### All cells

```{r}
fdr_cutoff <- 0.2
diffexp <- rbind(all_ac$toptables, all_aa$toptables) %>%
  dplyr::rename(cell = cluster_id,
                adj.P.Val = p_adj.loc)

logfc = diffexp %>% 
  dplyr::select(gene, cell, logFC, cc, condition) %>% 
  spread(gene, logFC, fill = 0)
logfc2 <- as.matrix(logfc[, -c(1:4)])
rownames(logfc2) <- logfc$cell

cctype <- logfc$cc
cctype[logfc$cc == "All cells"] <- logfc$cell[logfc$cc == "All cells"]
cctype[cctype == "CD4 T cells"] <- "T cells"
cctype[cctype == "CD8 T cells"] <- "T cells"
cctype[cctype == "NK cells"] <- "T cells"
cctype[cctype == "MNP"] <- "MNP cells"

fdr = diffexp %>% 
  dplyr::select(gene, cell, adj.P.Val, cc, condition) %>% 
  spread(gene, adj.P.Val, fill = 1)
fdr2 <- fdr[, -c(1:4)]

row_ha <- rowAnnotation(cells=cctype, col = list(
  cells = c("B cells"="#999999", "Epithelial cells"="#009E73", 
            "Mast cells"="#F0E442", "MNP cells"="#D55E00", "T cells"="#56B4E9")
))
logfc$condition[logfc$condition == "AA"] <- "Allergic Asthma"
logfc$condition[logfc$condition == "AC"] <- "Allergic Controls"

ht_allcells <- Heatmap(logfc2, right_annotation = row_ha,
        row_split = logfc$condition,
        cluster_rows = FALSE,
        border = TRUE, name="logFC",
    cell_fun = function(j, i, x, y, width, height, fill) {
        if(fdr2[i, j] < fdr_cutoff)
            grid.text("*", x, y, gp = gpar(fontsize = 20))
})
ht_allcells
ht_allcells <- grid.grabExpr(draw(ht_allcells, merge_legends = FALSE))
ggsave(here("results", "all_cells_heatmap.pdf"), 
       plot = ht_allcells, width = 18, height = 5)

## ncells
ncells_all <- data@meta.data %>% 
  dplyr::group_by(group, condition, cluster) %>% 
  summarise(Freq = n()) %>% 
  mutate(group_condition = paste(group, condition, sep="_")) %>% 
  dplyr::filter(condition != "Dil") %>% 
  ungroup() %>% 
  dplyr::select(-c("group", "condition")) %>% 
  spread(group_condition, Freq) %>% 
  dplyr::rename(annotation = "cluster")

p1 <- scale(ncells_all[,-1], center = FALSE,
      scale = colSums(ncells_all[,-1])) %>% 
  as.data.frame() %>% 
  mutate(annotation = as.character(ncells_all$annotation)) %>% 
  gather(cond, prop, -annotation) %>% 
  mutate(cond = factor(cond, c("AC_Bln", "AC_Ag", "AA_Bln", "AA_Ag"))) %>% 
  ggplot(aes(x = cond, y = prop, fill = annotation)) +
  geom_bar(stat="identity", color = "black") +
  xlab("condition") +
  ylab("Cell proportions")
ggsave(here::here("results", "cell_prop_allcells.png"), p1)

p1

## nsig genes
nsig_all <- diffexp %>% 
  mutate(fc = ifelse(logFC > 0, "up", "down")) %>% 
  dplyr::group_by(condition, cell, fc) %>% 
  dplyr::filter(adj.P.Val < fdr_cutoff) %>% 
  summarise(n = n()) %>% 
  mutate(cond_fc = paste(condition,fc, sep="_")) %>% 
  ungroup() %>% 
  dplyr::select(-c(condition, fc)) %>% 
  spread(cond_fc, n, fill= 0)

knitr::kable(nsig_all)
```

### MNP cells

```{r}

diffexp <- rbind(mnp_aa$toptables, mnp_ac$toptables) %>% 
  dplyr::rename(cell = cluster_id,
                adj.P.Val = p_adj.loc)

logfc = diffexp %>%  
  dplyr::select(gene, cell, logFC, cc, condition) %>% 
  spread(gene, logFC, fill = 0)
logfc2 <- as.matrix(logfc[, -c(1:4)])
rownames(logfc2) <- logfc$cell

cctype <- logfc$cell

fdr = diffexp %>%  
  dplyr::select(gene, cell, adj.P.Val, cc, condition) %>% 
  spread(gene, adj.P.Val, fill = 1)
fdr2 <- fdr[, -c(1:4)]

row_ha <- rowAnnotation(cells=cctype)
logfc$condition[logfc$condition == "AA"] <- "Allergic Asthma"
logfc$condition[logfc$condition == "AC"] <- "Allergic Controls"


mnp_cells <- Heatmap(logfc2, right_annotation = row_ha,
        row_split = logfc$condition,
        border = TRUE, name="logFC",
    cell_fun = function(j, i, x, y, width, height, fill) {
        if(fdr2[i, j] < fdr_cutoff)
            grid.text("*", x, y, gp = gpar(fontsize = 20))
})
mnp_cells
mnp_cells <- grid.grabExpr(draw(mnp_cells, merge_legends = FALSE))
ggsave(here::here("results", "mnp_cells_heatmap.pdf"), 
       plot = mnp_cells, width = 18, height = 5)

## ncells
ncells_mnp <- mnp@meta.data %>% 
  dplyr::group_by(group, condition, annotation) %>% 
  summarise(Freq = n()) %>% 
  mutate(group_condition = paste(group, condition, sep="_")) %>% 
  dplyr::filter(condition != "Dil") %>% 
  ungroup() %>% 
  dplyr::select(-c("group", "condition")) %>% 
  spread(group_condition, Freq)

p2 <- scale(ncells_mnp[,-1], center = FALSE,
      scale = colSums(ncells_mnp[,-1])) %>% 
  as.data.frame() %>% 
  mutate(annotation = as.character(ncells_mnp$annotation)) %>% 
  gather(cond, prop, -annotation) %>% 
  mutate(cond = factor(cond, c("AC_Bln", "AC_Ag", "AA_Bln", "AA_Ag"))) %>% 
  ggplot(aes(x = cond, y = prop, fill = annotation)) +
  geom_bar(stat="identity", color = "black") +
  xlab("condition") +
  ylab("Cell proportions")
ggsave(here::here("results", "cell_prop_mnp.png"), p2, height = 5)

p2

## nsig genes
nsig_mnp <- diffexp %>% 
  mutate(fc = ifelse(logFC > 0, "up", "down")) %>% 
  dplyr::group_by(condition, cell, fc) %>% 
  dplyr::filter(adj.P.Val < fdr_cutoff) %>% 
  summarise(n = n()) %>% 
  mutate(cond_fc = paste(condition,fc, sep="_")) %>% 
  ungroup() %>% 
  dplyr::select(-c(condition, fc)) %>% 
  spread(cond_fc, n, fill= 0)
knitr::kable(nsig_mnp)
```

### T cells

```{r}

diffexp <- rbind(tcell_ac$toptables, tcell_aa$toptables) %>% 
  dplyr::rename(cell = cluster_id,
                adj.P.Val = p_adj.loc)

logfc = diffexp %>% 
  dplyr::select(gene, cell, logFC, cc, condition) %>% 
  spread(gene, logFC, fill = 0)
logfc2 <- as.matrix(logfc[, -c(1:4)])
rownames(logfc2) <- logfc$cell

cctype <- logfc$cell

fdr = diffexp %>% 
  dplyr::select(gene, cell, adj.P.Val, cc, condition) %>% 
  spread(gene, adj.P.Val, fill = 1)
fdr2 <- fdr[, -c(1:4)]
# fdr2[is.na(fdr2)] <- 1

row_ha <- rowAnnotation(cells=cctype)
logfc$condition[logfc$condition == "AA"] <- "Allergic Asthma"
logfc$condition[logfc$condition == "AC"] <- "Allergic Controls"

t_cells <- Heatmap(logfc2, right_annotation = row_ha, 
        row_split = logfc$condition,
        border = TRUE, name="logFC",
    cell_fun = function(j, i, x, y, width, height, fill) {
        if(fdr2[i, j] < fdr_cutoff)
            grid.text("*", x, y, gp = gpar(fontsize = 20))
})
t_cells
t_cells <- grid.grabExpr(draw(t_cells, merge_legends = FALSE))
ggsave(here::here("results", "t_cells_heatmap.pdf"), 
       plot = t_cells, width = 18, height = 5)

## ncells
ncells_tcells <- tcell@meta.data %>% 
  dplyr::group_by(group, condition, annotation) %>% 
  summarise(Freq = n()) %>% 
  mutate(group_condition = paste(group, condition, sep="_")) %>% 
  dplyr::filter(condition != "Dil") %>% 
  ungroup() %>% 
  dplyr::select(-c("group", "condition")) %>% 
  spread(group_condition, Freq)

p3 <- scale(ncells_tcells[,-1], center = FALSE,
      scale = colSums(ncells_tcells[,-1])) %>% 
  as.data.frame() %>% 
  mutate(annotation = as.character(ncells_tcells$annotation)) %>% 
  gather(cond, prop, -annotation) %>% 
  mutate(cond = factor(cond, c("AC_Bln", "AC_Ag", "AA_Bln", "AA_Ag"))) %>% 
  ggplot(aes(x = cond, y = prop, fill = annotation)) +
  geom_bar(stat="identity", color = "black") +
  xlab("condition") +
  ylab("Cell proportions")
ggsave(here::here("results", "cell_prop_tcells.png"), p3, height = 5)
p4 <- cowplot::plot_grid(p1, p2, p3, labels = c("A", "B", "C"), ncol = 1)
ggsave(here::here("results", "cell_prop_combined.png"), p4, height = 12)

p4

## nsig genes
nsig_tcells <- diffexp %>% 
  mutate(fc = ifelse(logFC > 0, "up", "down")) %>% 
  dplyr::group_by(condition, cell, fc) %>% 
  dplyr::filter(adj.P.Val < fdr_cutoff) %>% 
  summarise(n = n()) %>% 
  mutate(cond_fc = paste(condition,fc, sep="_")) %>% 
  ungroup() %>% 
  dplyr::select(-c(condition, fc)) %>% 
  spread(cond_fc, n, fill= 0)

knitr::kable(nsig_tcells)
```

## number of cells

```{r}
ncells <- cbind(cellclass = c(as.character(ncells_all$annotation), rep("MNP", nrow(ncells_mnp)), rep("T cells", nrow(ncells_tcells))), 
                rbind(ncells_all, ncells_mnp, ncells_tcells))
write.csv(ncells[, c("cellclass", "annotation", "AC_Bln", "AC_Ag", "AA_Bln", "AA_Ag")], here::here("results", "ncells.csv"))

knitr::kable(ncells)
```

## number of significant gene-transcripts

```{r}
nsig <- cbind(cellclass = c(as.character(nsig_all$cell), 
                       rep("MNP", nrow(nsig_mnp)), 
                       rep("T cells", nrow(nsig_tcells))), 
                rbind(nsig_all, nsig_mnp, nsig_tcells))
write.csv(nsig[, c("cellclass", "cell", "AC_up", "AC_down","AA_up", "AA_down")], here::here("results", "nsig.csv"))

knitr::kable(nsig)
```

# sessionInfo

```{r}
sessionInfo()
```
