---
title: "Untitled"
author: "Amrit Singh"
date: "2025-04-18"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

library(anndata)
library(Seurat)
library(tidyverse)
library(SingleCellExperiment)
library(muscat)
library(scater)
library(limma)
library(ComplexHeatmap)
library(fgsea)
library(parallel)
library(dplyr)
source(here::here("src", "utils.R"))

latephase_exacer_biomarkers <- readRDS(here::here("results", "latephase_exacer_biomarkers.rds"))

```

## import data

```{r}
options(future.globals.maxSize = 16000 * 1024^2)
scperturb <- read_h5ad(here::here("data", "preprocessedData", "GSE279945_sc_counts_processed.h5ad"))
scperturb <- CreateSeuratObject(counts = t(as.matrix(scperturb$X)), 
                                meta.data = scperturb$obs)
sce <- SingleCellExperiment(
  assays = list(counts = scperturb@assays$RNA$counts),
  colData = scperturb@meta.data
)
```

## UMAP

```{r}
set.seed(123)  # For reproducibility
sampled_cells <- sample(colnames(scperturb), 100000)

# 2. Subset the Seurat object to include only the sampled cells
seurat_subset <- subset(scperturb, cells = sampled_cells)

seurat_subset <- NormalizeData(seurat_subset,
                      normalization.method = "LogNormalize") %>% 
  FindVariableFeatures(selection.method = "vst", 
                       nfeatures = 2000) %>% 
  ScaleData(features = rownames(seurat_subset)) %>% 
  RunPCA() %>% 
  RunUMAP(dims = 1:50, reduction.name = 'umap.rna', 
          reduction.key = 'rnaUMAP_')

drug_umap <- DimPlot(seurat_subset, reduction = "umap.rna", label = TRUE, 
                   label.size = 2.5, repel = TRUE, group.by = "cell_type_orig") + 
  ggtitle("RNA - All cells") +
  xlab("UMAP 1") +
  ylab("UMAP 2")
ggsave(here::here("results", "drug_umap_scrnaseq.png"), drug_umap, height = 4, width = 6)
drug_umap
```


## run differential expression analysis for each drug

```{r}
drug_result <- sc_drugde(sce)
```

```{r}
gene_sig <- readRDS(here::here("results", "sc_exacer_aa_toptable.rds")) 

keep_cc <- c("NK cells", "MNP", "CD8 T cells", "CD4 T cells", "B cells")
lar_biomarkers = gene_sig %>% 
  filter(p_adj.loc < 0.2, cluster_id %in% keep_cc) %>% 
  mutate(FC = ifelse(logFC > 0, "up", "down")) %>% 
  ungroup() %>% 
  dplyr::select(cluster_id, FC, gene) %>% 
  group_by(cluster_id, FC) %>% 
  nest() %>% 
  dplyr::rename(geneset = data) %>% 
  dplyr::rename(cell_type = cluster_id)
```

## ENRICHMENT analysis

```{r}
drug_de <- drug_result$toptables %>% 
  mutate(rank = logFC * log10(P.Value)) %>% 
  dplyr::select(sm_name, gene, rank, cc) %>% 
  spread(gene, rank)
drug_de$cc[drug_de$cc == "T cells CD4+"] <- "CD4 T cells"
drug_de$cc[drug_de$cc == "T cells CD8+"] <- "CD8 T cells"
drug_de$cc[drug_de$cc == "Myeloid cells"] <- "MNP"
drug_de[is.na(drug_de)] <- 0

a = drug_de %>% 
  group_by(sm_name) %>% 
  nest()
my_list <- split(a$data, a$sm_name)

my_fun <- function(x) {
  row_list <- lapply(split(x[[1]][,-1], seq(nrow(x[[1]]))), function(i){
      ranks <- as.numeric(i)
      names(ranks) <- colnames(i)
      ranks
    })
  names(row_list) <- x[[1]]$cc
  
  lapply(names(row_list), function(cell){
    cc <- subset(lar_biomarkers, cell_type == cell)
    genesets <- lapply(cc$geneset, function(i) pull(i, "gene"))
    names(genesets) <- paste(cc$cell_type, cc$FC, sep="_")
    tryCatch(fgsea(genesets, row_list[[cell]]) %>% 
      mutate(cell = cell),
      error = function(e) data.frame("pathway"= NA, "pval" = NA, 
                                     "padj"= NA, "log2err" = NA,
                                     "ES" = NA, "NES" = NA,
                                     "size" = NA, "leadingEdge" = NA,
                                     "cell" = NA))
  }) %>% 
    do.call(rbind, .)
}

cl <- makeCluster(10)
clusterEvalQ(cl, library(fgsea)) 
clusterEvalQ(cl, library(tidyverse)) 
clusterExport(cl, varlist = c("lar_biomarkers"))
result <- parLapply(cl, my_list, my_fun) 
stopCluster(cl)

res <- lapply(names(result), function(i){
  result[[i]] %>% 
    mutate(sm_name = i)
}) %>% 
  do.call(rbind, .)
```

## connectivity score

```{r}
cmap_diff = res %>% 
  dplyr::select(sm_name, pathway, NES) %>% 
  separate(col="pathway", into=c("cell", "direction"), sep="_") %>% 
  spread(direction, NES, fill = 0) %>% 
  group_by(cell) %>% 
  mutate(sign_down = sign(down),
         sign_up = sign(up)) %>% 
  filter(sign_down != sign_up,
         sign_up != 0,
         sign_down != 0) %>% 
  mutate(score = up-down)

## drug by cell
cmap_diff_cell = cmap_diff %>% 
  group_by(cell) %>% 
  arrange(score) %>% 
  dplyr::select(sm_name, cell, score) %>% 
  mutate(method="CMap") %>% 
  group_by(method, cell) %>% 
  mutate(rank = rank(score),
         rank = rank/max(rank))
cmap_diff_cell$sm_name <- as.character(cmap_diff_cell$sm_name)
cmap_diff_cell$sm_name[grep("Isopropyl", cmap_diff_cell$sm_name)] <- "PMP-5A"

df_summarized <- cmap_diff_cell %>%
  group_by(sm_name) %>%
  summarise(total = sum(score)) %>% 
  filter(abs(total) > 3) %>% 
  arrange(total)

p <- cmap_diff_cell %>% 
  filter(sm_name %in% as.character(df_summarized$sm_name)) %>% 
  mutate(sm_name = factor(as.character(sm_name),
                                 rev(as.character(df_summarized$sm_name)))) %>% 
  ggplot(aes(y = sm_name, x = score, 
             fill = cell)) +
  geom_bar(stat="identity", color = "black") +
  ylab("Drug") +
  xlab("Connectivity Score") +
  theme_classic() +
  theme(legend.position = c(0.9, 0.9))


## overall drug
cmap_diff_all = cmap_diff_cell %>% 
  dplyr::group_by(method, sm_name) %>% 
  summarise(score = mean(score)) %>% 
  mutate(rank = rank(score),
         rank = rank/max(rank)) %>% 
  arrange(rank)


## plot fold-changes of top 4 drugs
plot_drug_fc = function(drug_toptable, disease_toptable,
                        drug="Mometasone Furoate"){
  p <- drug_toptable %>% 
    filter(sm_name == drug) %>% 
    dplyr::select(sm_name, pathway, cell, NES, leadingEdge) %>% 
    unnest(leadingEdge) %>% 
    dplyr::rename(drug_fc = NES, gene = leadingEdge) %>% 
    inner_join(disease_toptable %>% 
                 ungroup() %>% 
                 dplyr::select(gene, cluster_id, logFC) %>% 
                 dplyr::rename(cell = cluster_id, asthma_fc = logFC), 
               by=c("cell", "gene")) %>% 
    ggplot(aes(x = asthma_fc, y = drug_fc, color = cell)) +
    geom_hline(yintercept = 0, linetype = "dashed", color = "grey") +
    geom_vline(xintercept = 0, linetype = "dashed",  color = "grey") +
    geom_point() +
    ggrepel::geom_text_repel(aes(label = gene), 
                              size = 3,
                             fontface = "italic",
                             max.overlaps = 50) +
    xlab("Fold-change (Pre vs. Post allergen challenge)") +
    ylab(paste0("Fold-change \n (DMSO vs. ", drug, ")")) +
    theme_classic() +
    theme(legend.position = "none") +
    xlim(c(-0.75, 0.75)) +
    ylim(c(-2, 2))
  p
}

p_list <- lapply(df_summarized$sm_name[1:3], function(drug){
  plot_drug_fc(drug_toptable = res, disease_toptable = gene_sig,
                        drug=drug)
})
p2 <- cowplot::plot_grid(plotlist = p_list, ncol=1)

p3 <- cowplot::plot_grid(p, p2, ncol = 2, rel_widths = c(2.5, 1),
                         labels = c("A", "B"))
ggsave(here::here("results", "sc_cmap.png"), p3, height = 10, width = 13)


```


